src/include/gurt/common.h

```diff
@@ -65,14 +65,66 @@ extern "C" {
 
 /* memory allocating macros */
 void  d_free(void *);
-void *d_calloc(size_t, size_t);
-void *d_malloc(size_t);
-void *d_realloc(void *, size_t);
+void *d_calloc(int tag, size_t, size_t);
+void *d_malloc(int tag, size_t);
+void *d_realloc(int tag, void *, size_t);
 char *d_strndup(const char *s, size_t n);
 int d_asprintf(char **strp, const char *fmt, ...);
-void *d_aligned_alloc(size_t alignment, size_t size);
+void *d_aligned_alloc(int tag, size_t alignment, size_t size);
 char *d_realpath(const char *path, char *resolved_path);
 
+/** Allocation tag, please make sure dm_counters are declared in the same order. */
+enum dm_tag {
+	M_OTHER,		/* all the untagged */
+	M_STR,			/* strdup, asprintf... */
+	M_RANK,			/* rank list allocation */
+	M_CRT,			/* default cart tag */
+	M_CRT_RPC,		/* CaRT RPC */
+	M_CRT_IV,		/* IV framework in cart */
+	M_SWIM,			/* SWIM */
+	M_CSUM,			/* checksum */
+	M_PROP,			/* properties */
+	M_TSE,			/* task engine */
+	M_PL,			/* placement */
+	M_IV,			/* DAOS level IV */
+	M_SCHED,		/* ABT scheduler */
+	M_IO,			/* common I/O code */
+	M_IO_ARG,		/* SGL, IOV... */
+	M_BIO,			/* BIO */
+	M_VEA,			/* VEA */
+	M_AGG,			/* common aggregation */
+	M_RECOV,		/* common rebuild */
+	M_EC,			/* common EC */
+	M_EC_AGG,		/* EC aggregation */
+	M_EC_RECOV,		/* EC recovery */
+	M_POOL,			/* Pool, cache or data structures */
+	M_CONT,			/* container, cache or data structures */
+	M_OBJ,			/* object, shard, cache... */
+	M_DTX,			/* distributed transaction */
+	M_VOS,			/* common VOS */
+	M_VOS_TS,		/* VOS timestamp cache */
+	M_VOS_LRU,		/* VOS LRU */
+	M_VOS_DTX,		/* DTX in VOS */
+	M_RDB,			/* Raft DB */
+	M_TAG_MAX,		/* the last */
+};
+
+struct dm_tls_counter;
+
+/* 32 bytes header in memory-debugging mode */
+struct dm_header {
+	struct dm_tls_counter	*mh_counter;
+	/** real address of the buffer */
+	void			*mh_addr;
+	int64_t			 mh_size;
+	int32_t			 mh_alignment;
+	int32_t			 mh_tag;
```

Comment:
I wonder if you can define such macro in each file like D_ALLOC_TAG

for example

#define D_ALLOC_TAG          CLI_OBJ_TAG  


And define CLI_OBJ_TAG in common.h, so each file will have identify tag, and plus with the line number suggested below, so each dm_header will have identify place.



And also you do not need change D_ALLOC and D_ALLOC_PTR interface,  since you can get D_ALLOC_TAG directly  i.e.

D_ALLOC(ptr, size)   D_ALLOC_CORE(ptr, size, D_ALLOC_TAG).

This should simplify the patch a lot?

Reply:
The reason I didn't do that is because we do use different tags in the same module, e.g. object module uses M_IO, M_OBJ, M_RECOV, M_CSUM, M_EC, M_IO_ARG, .... it's hard to use a global one.

Reply:
hmm, IMHO, the counter does not have to be purpose,  I think per module (or even per file) would be good enough for us?  Actually per module (or file) would be easier to debug compared with per purpose here, IMHO.

Reply:
seems not very easy to correctly use this kind of tags. If someone use mismatch tags for alloc/free in any future's patch (note that one alloc may correspond to different places' free) then will cause all the counters be not usable and cause trouble when debugging it.